File: frontend/src/domain/researchPreparation.ts
/**
 * Stable frontend vocabulary for the review-gated research preparation flow.
 * No preparation action implies provider execution; authorization is the only
 * command that can materialize its already-reviewed plan into a research run.
 */

export type ResearchPreparationStatus =
  | "preparing"
  | "awaiting_claim_review"
  | "awaiting_protocol_confirmation"
  | "awaiting_plan_authorization"
  | "recoverable_failure"
  | "authorized";

export type ResearchPreparationSystemStepState =
  | "queued"
  | "running"
  | "succeeded"
  | "retrying"
  | "failed"
  | "stale";

export type ResearchPreparationReviewStepState =
  | "locked"
  | "awaiting_review"
  | "confirmed"
  | "stale";

export type ResearchPreparationArtifactState = "current" | "stale" | "superseded";

export type ResearchPreparationEventStep =
  | "parse_claims"
  | "draft_protocol"
  | "draft_evidence_plan";

export interface ResearchPreparationInitialMaterial {
  documentVersionId: string;
  title: string | null;
  parseState: string;
}

export interface ResearchPreparationProgress {
  completedSteps: number;
  totalSteps: number;
  currentStep: ResearchPreparationEventStep | null;
  failedStep: ResearchPreparationEventStep | null;
}

export interface ResearchPreparationSystemStep {
  state: ResearchPreparationSystemStepState;
  artifactSequence: number | null;
}

export interface ResearchPreparationReviewStep {
  state: ResearchPreparationReviewStepState;
}

export interface ResearchPreparationStages<T> {
  candidateClaims: T;
  protocol: T;
  evidencePlan: T;
}

export interface ResearchPreparationArtifact {
  sequence: number;
  state: ResearchPreparationArtifactState;
  payload: Record<string, unknown>;
  contextFingerprint: string | null;
  /** Source-use policy prevents this artifact from being displayed or approved. */
  displayWithheld: boolean;
}

export interface ResearchPreparation {
  caseId: string;
  caseTitle?: string | null;
  initialMaterial?: ResearchPreparationInitialMaterial | null;
  progress?: ResearchPreparationProgress | null;
  revision: number;
  status: ResearchPreparationStatus;
  researchRunId: string | null;
  system: ResearchPreparationStages<ResearchPreparationSystemStep>;
  review: ResearchPreparationStages<ResearchPreparationReviewStep>;
  nextAttemptAt: string | null;
  lastErrorMessage: string | null;
  artifacts: ResearchPreparationStages<ResearchPreparationArtifact | null>;
  authorizedEvidencePlan: Record<string, unknown> | null;
}

export interface ResearchPreparationEvent {
  seq: number;
  type: string;
  step: ResearchPreparationEventStep | null;
  message: string | null;
  detail: Record<string, unknown> | null;
  createdAt: string;
}

export interface ResearchPreparationEventsPage {
  items: ResearchPreparationEvent[];
  nextAfterSeq: number | null;
}

export interface ConfirmResearchPreparationClaimsInput {
  caseId: string;
  revision: number;
  actor: string;
  decisions: Array<{
    candidateId: string;
    outcome: "confirmed" | "modified" | "rejected";
    reason: string;
    normalizedText?: string | null;
  }>;
}

export interface ConfirmResearchPreparationProtocolInput {
  caseId: string;
  revision: number;
  actor: string;
  draftSequence: number;
  edits?: Record<string, unknown>;
}

export interface RetryResearchPreparationInput {
  caseId: string;
  revision: number;
  actor: string;
}

export interface AuthorizeResearchPreparationInput {
  caseId: string;
  revision: number;
  actor: string;
  planSequence: number;
  idempotencyKey: string;
}

/** A revision/idempotency conflict that UI can handle without parsing text. */
export class ConflictError extends Error {
  readonly status = 409;
  constructor(message = "当前准备版本已变化，请刷新后重试。") {
    super(message);
    this.name = "ConflictError";
  }
}

File: backend/app/schemas/v1/research_preparation.py
"""Public, deliberately redacted research-preparation API contracts."""
from __future__ import annotations

import uuid
from typing import Any, Literal
from pydantic import Field
from app.schemas.v1.common import V1Model

class ClaimDecisionDTO(V1Model):
    candidate_id: uuid.UUID
    outcome: Literal["confirmed", "modified", "rejected"]
    reason: str = Field(min_length=1, max_length=2000)
    normalized_text: str | None = None
class ConfirmClaimsRequest(V1Model):
    revision: int = Field(ge=1)
    actor: str = Field(min_length=1, max_length=128)
    decisions: list[ClaimDecisionDTO] = Field(min_length=1)
class ConfirmProtocolRequest(V1Model):
    revision: int = Field(ge=1)
    actor: str = Field(min_length=1, max_length=128)
    draft_sequence: int = Field(ge=1)
    edits: dict[str, object] = Field(default_factory=dict)
class AuthorizeEvidencePlanRequest(V1Model):
    revision: int = Field(ge=1)
    actor: str = Field(min_length=1, max_length=128)
    plan_sequence: int = Field(ge=1)
    idempotency_key: str = Field(min_length=1, max_length=256)
class RetryResearchPreparationRequest(V1Model):
    revision: int = Field(ge=1)
    actor: str = Field(min_length=1, max_length=128)
class PreparationStepDTO(V1Model):
    state: str
    review_state: str | None = None
    artifact_sequence: int | None = None
class PreparationArtifactDTO(V1Model):
    sequence: int
    payload: dict[str, Any]
    state: str
    context_fingerprint: str | None = None
class PreparationInitialMaterialDTO(V1Model):
    document_version_id: uuid.UUID
    title: str | None = None
    parse_state: str
class PreparationProgressDTO(V1Model):
    completed_steps: int = Field(ge=0, le=3)
    total_steps: int = Field(ge=1, le=3)
    current_step: str | None = None
    failed_step: str | None = None
class ResearchPreparationDTO(V1Model):
    case_id: uuid.UUID
    case_title: str
    initial_material: PreparationInitialMaterialDTO | None = None
    progress: PreparationProgressDTO
    revision: int
    status: str
    research_run_id: uuid.UUID | None = None
    system: dict[str, PreparationStepDTO]
    review: dict[str, PreparationStepDTO]
    next_attempt_at: str | None = None
    last_error_message: str | None = None
    artifacts: dict[str, PreparationArtifactDTO | None]
    authorized_evidence_plan: dict[str, Any] | None = None
class ResearchPreparationEventDTO(V1Model):
    seq: int
    type: str
    step: str | None = None
    message: str | None = None
    detail: dict[str, Any] = Field(default_factory=dict)
    created_at: str
class ResearchPreparationEventsResponse(V1Model):
    items: list[ResearchPreparationEventDTO]
    next_after_seq: int | None = None
